### PR TITLE
Feat/inmemory cache

### PR DESCRIPTION
**Pull request by Filipki**
Branch was checked out from master 
¯\_(ツ)_/¯

### CODE REVIEW

File: F1WM.UnitTests/Controllers/RSSControllerTests.cs [hunk 1]
```diff
@@ -60,7 +60,7 @@ public async Task ShouldReturnCachedRSS()
 
 			serviceMock.Verify(s => s.GetFeed(null), Times.Never);
 			cachingServiceMock.Verify(c => c.Get<SyndicationFeed>(It.IsAny<string>()), Times.Once);
-			cachingServiceMock.Verify(c => c.Set(It.IsAny<string>(), It.IsAny<SyndicationFeed>(), It.IsAny<MemoryCacheEntryOptions>()), Times.Never);
+			cachingServiceMock.Verify(c => c.Set(It.IsAny<string>(), It.IsAny<SyndicationFeed>(), TimeSpan.Zero), Times.Never);
```
Comment (pevel) [inline F1WM.UnitTests/Controllers/RSSControllerTests.cs:63]: We should make sure that `Set` isn't called with *any* param, so `TimeSpan.Zero` should be changed to fitting `It.IsAny` setup. The same applies to all new assertions that check for `Times.Never` and have specific `TimeSpan` params

File: F1WM.UnitTests/Controllers/RacesControllerTests.cs [hunk 2]
```diff
@@ -173,5 +177,69 @@ public async Task ShouldReturn404IfDriversStandingsAfterRaceNotFound()
...
+			DateTime? requestParam = null;
+			var cacheKey = $"{CacheKeys.NextRace}_{requestParam}"; ;
+
+			cachingServiceMock.Setup(c => c.TryGetCacheValue<NextRaceSummary>(cacheKey)).Returns(new NextRaceSummary());
+
+			var result = await controller.GetNextRace(requestParam);
+
+			racesServiceMock.Verify(s => s.GetNextRace(requestParam), Times.Never);
+			cachingServiceMock.Verify(c => c.TryGetCacheValue<NextRaceSummary>(cacheKey), Times.Once);
+			cachingServiceMock.Verify(c => c.Set(cacheKey, It.IsAny<NextRaceSummary>(), TimeSpan.FromDays(1)), Times.Never);
```
Comment (pevel) [inline F1WM.UnitTests/Controllers/RacesControllerTests.cs:193]: We should make sure that `Set` isn't called with any param, so concrete `TimeSpan` should be changed to fitting `It.IsAny` setup

File: F1WM.UnitTests/Controllers/RacesControllerTests.cs [hunk 3]
```diff
@@ -173,5 +177,69 @@ public async Task ShouldReturn404IfDriversStandingsAfterRaceNotFound()
...
+			DateTime? requestParam = null;
+			var cacheKey = $"{CacheKeys.LastRace}_{requestParam}"; ;
+
+			cachingServiceMock.Setup(c => c.TryGetCacheValue<LastRaceSummary>(cacheKey)).Returns(new LastRaceSummary());
+
+			var result = await controller.GetLastRace(requestParam);
+
+			racesServiceMock.Verify(s => s.GetLastRace(requestParam), Times.Never);
+			cachingServiceMock.Verify(c => c.TryGetCacheValue<LastRaceSummary>(cacheKey), Times.Once);
+			cachingServiceMock.Verify(c => c.Set(cacheKey, It.IsAny<LastRaceSummary>(), TimeSpan.FromDays(1)), Times.Never);
```
Comment (pevel) [inline F1WM.UnitTests/Controllers/RacesControllerTests.cs:209]: We should make sure that `Set` isn't called with any param, so concrete `TimeSpan` should be changed to fitting `It.IsAny` setup

File: F1WM.UnitTests/Controllers/RacesControllerTests.cs [hunk 4]
```diff
@@ -173,5 +177,69 @@ public async Task ShouldReturn404IfDriversStandingsAfterRaceNotFound()
 			standingsServiceMock.Verify(s => s.GetDriversStandingsAfterRace(raceId), Times.Once);
 			Assert.IsType<NotFoundResult>(result.Result);
 		}
+
+		[Fact]
+		public async Task ShouldReturnCachedNextRace()
+		{
+			DateTime? requestParam = null;
+			var cacheKey = $"{CacheKeys.NextRace}_{requestParam}"; ;
```
Comment (pevel) [inline F1WM.UnitTests/Controllers/RacesControllerTests.cs:185]: Wild double semicolon

File: F1WM.UnitTests/Controllers/RacesControllerTests.cs [hunk 5]
```diff
@@ -173,5 +177,69 @@ public async Task ShouldReturn404IfDriversStandingsAfterRaceNotFound()
...
+			cachingServiceMock.Verify(c => c.TryGetCacheValue<NextRaceSummary>(cacheKey), Times.Once);
+			cachingServiceMock.Verify(c => c.Set(cacheKey, It.IsAny<NextRaceSummary>(), TimeSpan.FromDays(1)), Times.Never);
+			var okResult = Assert.IsType<OkObjectResult>(result.Result);
+		}
+
+		[Fact]
+		public async Task ShouldReturnCachedLastRace()
+		{
+			DateTime? requestParam = null;
+			var cacheKey = $"{CacheKeys.LastRace}_{requestParam}"; ;
```
Comment (pevel) [inline F1WM.UnitTests/Controllers/RacesControllerTests.cs:201]: Wild double semicolon

File: F1WM.UnitTests/Controllers/StandingsControllerTests.cs [hunk 6]
```diff
@@ -54,5 +60,74 @@ public async Task ShouldReturnDriversStandingsBySeasonId()
...
+			int count = 100;
+			var cacheKey = $"{CacheKeys.ConstructorsStanding}_{seasonId}_{count}";
+
+			cachingServiceMock.Setup(c => c.TryGetCacheValue<ConstructorsStandings>(cacheKey)).Returns(new ConstructorsStandings());
+
+			var result = await controller.GetConstructorsStandings(seasonId, count);
+
+			serviceMock.Verify(s => s.GetConstructorsStandings(count, seasonId), Times.Never);
+			cachingServiceMock.Verify(c => c.TryGetCacheValue<ConstructorsStandings>(cacheKey), Times.Once);
+			cachingServiceMock.Verify(c => c.Set(cacheKey, It.IsAny<ConstructorsStandings>(), TimeSpan.FromDays(1)), Times.Never);
```
Comment (pevel) [inline F1WM.UnitTests/Controllers/StandingsControllerTests.cs:77]: We should make sure that `Set` isn't called with any param, so concrete `TimeSpan` should be changed to fitting `It.IsAny` setup

File: F1WM.UnitTests/Controllers/StandingsControllerTests.cs [hunk 7]
```diff
@@ -54,5 +60,74 @@ public async Task ShouldReturnDriversStandingsBySeasonId()
...
+			int count = 100;
+			var cacheKey = $"{CacheKeys.DriversStanding}_{seasonId}_{count}";
+
+			cachingServiceMock.Setup(c => c.TryGetCacheValue<DriversStandings>(cacheKey)).Returns(new DriversStandings());
+
+			var result = await controller.GetDriversStandings(seasonId, count);
+
+			serviceMock.Verify(s => s.GetDriversStandings(count, seasonId), Times.Never);
+			cachingServiceMock.Verify(c => c.TryGetCacheValue<DriversStandings>(cacheKey), Times.Once);
+			cachingServiceMock.Verify(c => c.Set(cacheKey, It.IsAny<DriversStandings>(), TimeSpan.FromDays(1)), Times.Never);
```
Comment (pevel) [inline F1WM.UnitTests/Controllers/StandingsControllerTests.cs:94]: We should make sure that `Set` isn't called with any param, so concrete `TimeSpan` should be changed to fitting `It.IsAny` setup

File: F1WM/Controllers/CachingController.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+	[Route("api/[controller]")]
+	public class CachingController : ControllerBase
+	{
+		private readonly ICachingService service;
+
+		[HttpDelete]
+		[ProducesResponseType(204)]
+		[ProducesResponseType(401)]
+		[ProducesResponseType(404)]
+		public ActionResult ClearCache()
```
Comment (pevel) [inline F1WM/Controllers/CachingController.cs:15]: Please add `[Authorize]` to this endpoint

File: F1WM/Controllers/NewsController.cs [hunk 9]
```diff
@@ -69,7 +105,13 @@ public async Task<IEnumerable<NewsTagCategory>> GetTagCategories()
 		[HttpGet("important")]
 		public async Task<IEnumerable<ImportantNewsSummary>> GetImportantNews()
 		{
-			return await service.GetImportantNews();
+			var responseData = cachingService.TryGetCacheValue<IEnumerable<ImportantNewsSummary>>(CacheKeys.ImportantNews.ToString());
+			if (responseData is null || responseData.Any())
+			{
+				responseData = await service.GetImportantNews();
+				cachingService.Set(CacheKeys.ImportantNews.ToString(), responseData, TimeSpan.FromDays(5));
```
Comment (pevel) [inline F1WM/Controllers/NewsController.cs:112]: Can we afford to make cache duration so long? News change quite frequently.

File: F1WM/Services/CachingService.cs [hunk 10]
```diff
@@ -1,25 +1,50 @@
...
+
+		public void Set<T>(string key, T value, TimeSpan absoluteExpirationRelativeToNow)
 		{
-			return cache.Set<T>(key, value, options);
+			var options = new MemoryCacheEntryOptions().SetPriority(CacheItemPriority.Normal).SetSize(1).SetAbsoluteExpiration(absoluteExpirationRelativeToNow);
+			options.AddExpirationToken(new CancellationChangeToken(_resetCacheToken.Token));
+			cache.Set<T>(key, value, options);
+		}
+
+		public void DisposeCache()
```
Comment (pevel) [inline F1WM/Services/CachingService.cs:34]: You might want to add methods for disposing specific cache keys (or groups of keys, using common `CancellationTokenSource` - [example](https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.1#cache-dependencies-1)). This way we wouldn't need to drop whole cache when only a portion of data is affected, e.g. right now we drop the cache when broadcasts info changes.

**Comment by pevel** [comment]
Please also make sure unit tests pass!

File: F1WM/Controllers/NewsController.cs (hunk 9 above)
Comment (lukaszgodula) [reply F1WM/Controllers/NewsController.cs:112]: For now we will clear cache on every edit in db, so it's safe. I know that this solution is not perfect, but our old CMS is quite not friendly as for adding new features :)

File: F1WM/Services/CachingService.cs (hunk 10 above)
Comment (lukaszgodula) [reply F1WM/Services/CachingService.cs:34]: That's good idea but I would create such endpoint in lets say second iteration of our caching. For now the only CRUD in API is Broadcasts, so gain would not be so great. But with another endpoints we can add possibility to add specific keys and handle cache in 100% legit manner :)

**Review by pevel — APPROVED** [review]

**Pull request merged**